### PR TITLE
[Enhancement](wal) Add wal space back pressure

### PR DESCRIPTION
**Pull request by Yukang-Lian**
## Proposed changes

Issue Number: close #xxx

The current WAL is deleted after the transaction completes. I added a WAL space back pressure mechanism to control WAL disk usage. The default value is 64M.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/common/config.cpp [hunk 1]
```diff
@@ -1118,9 +1118,12 @@ DEFINE_Bool(ignore_always_true_predicate_for_segment, "true");
 // Dir of default timezone files
 DEFINE_String(default_tzfiles_path, "${DORIS_HOME}/zoneinfo");
 
-// Max size(bytes) of group commit queues, used for mem back pressure.
+// Max size(bytes) of group commit queues, used for mem back pressure, defult 64M.
 DEFINE_Int32(group_commit_max_queue_size, "65536");
 
+// Max size(bytes) of wal disk using, used for disk space back pressure, default 64M.
+DEFINE_Int32(wal_max_disk_size, "65535");
```
Comment (gavinchou) [inline be/src/common/config.cpp:1125]: 65536 should be 64KiB?

Comment (Yukang-Lian) [reply be/src/common/config.cpp:1125]: Sorry, my math is not good :cold_sweat:

File: be/src/olap/wal_manager.cpp [hunk 2]
```diff
@@ -241,6 +251,12 @@ size_t WalManager::get_wal_table_size(const std::string& table_id) {
 Status WalManager::delete_wal(int64_t wal_id) {
     {
         std::lock_guard<std::shared_mutex> wrlock(_wal_lock);
+        *_all_wal_disk_bytes -= _wal_id_to_writer_map[wal_id]->disk_bytes();
+        _wal_id_to_writer_map[wal_id]->cv.notify_one();
+        _wal_id_to_writer_map.erase(wal_id);
+        if (_wal_id_to_writer_map.empty()) {
+            CHECK(*_all_wal_disk_bytes == 0);
```
Comment (gavinchou) [inline be/src/olap/wal_manager.cpp:258]: Use CHECK_EQ() instead.

File: be/src/olap/wal_writer.h [hunk 3]
```diff
@@ -17,33 +17,44 @@
...
+    doris::ConditionVariable cv;
 
 private:
+    static constexpr size_t MAX_WAL_WRITE_WAIT_TIME = 1000;
     std::string _file_name;
     io::FileWriterPtr _file_writer;
     int64_t _count;
     int64_t _batch;
+    size_t _disk_bytes;
+    std::shared_ptr<std::atomic_size_t> _all_wal_disk_bytes;
```
Comment (gavinchou) [inline be/src/olap/wal_writer.h:56]: Memory order can be `relaxed` for all operations of _all_wal_disk_bytes.

**Comment by Yukang-Lian** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Comment by gavinchou** [comment]
LGTM

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**